File: src/base58.ts
import { init, g, unwrap } from './init';
import type { Base58CheckDecodeResult } from './types';

/** Base58 encoding and decoding (with and without check). */
export const base58 = {
  /** Encode bytes (hex) to a base58 string.
   *  Calls Go: base58.Encode() from btcutil/base58. */
  async encode(hexData: string): Promise<string> {
    await init();
    return unwrap<string>(g().base58.encode(hexData));
  },

  /** Decode a base58 string to bytes (hex).
   *  Calls Go: base58.Decode() from btcutil/base58. */
  async decode(str: string): Promise<string> {
    await init();
    return unwrap<string>(g().base58.decode(str));
  },

  /** Encode bytes (hex) with a version byte and checksum.
   *  Calls Go: base58.CheckEncode() from btcutil/base58. */
  async checkEncode(hexData: string, version: number): Promise<string> {
    await init();
    return unwrap<string>(g().base58.checkEncode(hexData, version));
  },

  /** Decode a base58check string, verifying the checksum.
   *  Calls Go: base58.CheckDecode() from btcutil/base58. */
  async checkDecode(str: string): Promise<Base58CheckDecodeResult> {
    await init();
    return unwrap<Base58CheckDecodeResult>(g().base58.checkDecode(str));
  },
};

File: wasm-wrapper/base58.go
//go:build js && wasm

package main

import (
	"encoding/hex"
	"syscall/js"

	"github.com/btcsuite/btcd/btcutil/base58"
)

func base58Encode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexData"); e != nil {
		return e
	}
	b, e := hexDecode(args[0].String())
	if e != nil {
		return e
	}
	return okResult(base58.Encode(b))
}

func base58Decode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "str"); e != nil {
		return e
	}
	return okResult(hex.EncodeToString(base58.Decode(args[0].String())))
}

func base58CheckEncode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "hexData, version"); e != nil {
		return e
	}
	b, e := hexDecode(args[0].String())
	if e != nil {
		return e
	}
	version := byte(args[1].Int())
	return okResult(base58.CheckEncode(b, version))
}

func base58CheckDecode(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "str"); e != nil {
		return e
	}
	data, version, err := base58.CheckDecode(args[0].String())
	if err != nil {
		return errfResult("checkDecode failed: %s", err)
	}
	return okResult(map[string]any{
		"data":    hex.EncodeToString(data),
		"version": int(version),
	})
}
